### DISCUSSION TITLE
MacOS cannot successfully launch AIR apps with ANE native extensions from IntelliJ

### DISCUSSION THREAD

**Discussion by philpme** [airsdk/Adobe-Runtime-Support#2581] [Q&A] [answered]
We've come across an issue where we cannot successfully launch our app through IntelliJ (Ultimate 2023.1) on MacOS (Ventura 13.2.1). We've tested this with AIR 50.1 and 50.2. Our initial investigation has shown that the Flash/Flex IntelliJ plugin (latest 231.8109.90) is not correctly expanding the ANE into the temporary IntelliJ_ANE_unzipped folder. This leads to the MacOS Framework package becoming invalid with its internal symlinks replaced as regular files.  As a result ADL is no longer enabling the MacOS native code.

Looking at the source of the Flash/Flex plugin - the unzipANEFiles() method looks to have remained unchanged for at least the past five years, so it doesn't appear to be related to any change on the IntelliJ or plugin side, but likely a emerged shortcoming with either a later version of AIR or MacOS.
https://github.com/JetBrains/intellij-plugins/blob/master/flex/src/com/intellij/lang/javascript/flex/build/FlexCompilationUtils.java

We know the ANE package is fine, as the code run successfully when built into a captive runtime, which is then signed and notarized by Apple and then run on another machine.

Looking at the how ADL is launched by IntelliJ
`/Users/phil/Projects/flex-sdk-4.13-air50.2.2.3/bin/adl -profile extendedDesktop -extdir /private/var/folders/kk/j0rmf1p921j7htyrn8y_7g_c0000gp/T/IntelliJ_ANE_unzipped /Users/phil/Projects/project/client/flex/theapp/bin-debug/the-app.xml /Users/phil/Projects/project/client/flex/theapp/bin-debug -- 
`

When launched we get the following error

> [Fault] exception, information=ArgumentError: Error #3500: The extension context does not have a method with the name theMethodName.

We can see that the expanded ANE framework folder doesn't contain the expected symlinks
```
$ ls -la /private/var/folders/kk//j0rmf1p921j7htyrn8y_7g_c0000gp/T/IntelliJ_ANE_unzipped/TheANE64-1.0.0-225.ane/META-INF/ANE/MacOS-x86-64/TheANE.framework/
total 0
drwxr-xr-x@ 6 phil  staff   192 11 Apr 12:32 .
drwxr-xr-x@ 5 phil  staff   160 11 Apr 12:32 ..
-rw-r--r--@ 1 phil  staff    26 11 Apr 12:30 TheANE
-rw-r--r--@ 1 phil  staff    26 11 Apr 12:30 Resources
drwxr-xr-x@ 5 phil  staff   160 11 Apr 12:32 Versions
```

If instead we manually expand the ANE, and then list its contents
```
$ unzip TheANE64-1.0.0-225.ane
$ ls -la META-INF/ANE/MacOS-x86-64/TheANE.framework/
total 0
drwxr-xr-x  5 phil  staff  160 11 Apr 12:33 .
drwxr-xr-x  4 phil  staff  128 11 Apr 12:33 ..
lrwxrwxr-x  1 phil  staff   26 11 Apr 12:33 TheANE -> Versions/Current/TheANE
lrwxrwxr-x  1 phil  staff   26 11 Apr 12:33 Resources -> Versions/Current/Resources
drwxr-xr-x  4 phil  staff  128 11 Apr 12:33 Versions
```

We can see that the symlinks are maintained as expected.  Then if we instead manually launch ADL pointing to our expanded folder, the application launches correctly

 `/Users/phil/Projects/flex-sdk-4.13-air50.2.2.3/bin/adl -profile extendedDesktop -extdir /Users/phil/manualtest/ /Users/phil/Projects/project/client/flex/theapp/bin-debug/the-app.xml /Users/phil/Projects/project/client/flex/theapp/bin-debug -- 
`

Our guess is that either MacOS or AIR has tightened how native Framework extensions are loaded, and is now simply ignoring ANE when expanded by IntelliJ Flash/Flex plugin.  It would appear that this then needs a fix on the Flex/Flash plugin side.

But has anyone else come up with a workaround for this scenario until the ANE is correctly expanded in the IntelliJ_ANE_unzipped folder?

**Comment by marchbold** [comment, marked as answer]
We have experienced exactly the same issue, and as far as we can tell this is a change in macOS enforcement of what constitutes a valid framework. In the past folders with the correct structure have worked but now the must have the symlink to the Versions folder. Unfortunately java zip extraction doesn't support symlinks so the ANE expansion follows all the symlinks. 

The workaround is to get the temp folder that IntelliJ is using and manually extract the ANEs to this folder. IntelliJ normally uses the same temp folder for a session so you can do this after your first run and then debug your application as required.

As a complete solution we need to update the plugin. I have started working with Harman in setting up a development environment for the flash/flex IntelliJ plugin and we'll have to find an alternative way of extracting the ANEs and hopefully can get Jetbrains to push the update. Andrew has started initial conversation with them and sounds positive.

**Reply by philpme** [reply]
Hi Michael - has there been any progress on this? We are looking to upgrade our developers shortly, and it would be good to have this resolved in IntelliJ

**Reply by marchbold** [reply]
Unfortunately I haven't had any time to work on the intellij plugin. Spare time is a rare thing these days.

**Reply by philpme** [reply]
No stress - we'll fix internally and get a local build working and we'll raise a PR with the JetBrains repo

**Reply by marchbold** [reply]
Brilliant! That would be amazing if you have time. Let me know how it goes.

**Reply by philpme** [reply]
PR for proposed fix https://github.com/JetBrains/intellij-plugins/pull/898

It's quite simple, but as IJ have moved to Java 17, i believe the setup instructions for the flex-plugin are out of date. There are a number of compilation errors including

java: annotation @org.jetbrains.annotations.NotNull not applicable in this type context
form-bindings: Class to bind does not exist or is excluded from compilation

I gave up trying to get it to compile given the 100+ compilation errors right out of the box.

**Reply by marchbold** [reply]
Nice! Looks like a simple solution, hopefully Jetbrains can merge it in quickly.

**Comment by philpme** [comment]
Great news Michael. I had checked out the intellji-plugins and intellij-community repos to take a look what was going on. 

It appears that `https://github.com/JetBrains/intellij-community/blob/master/platform/util/src/com/intellij/util/io/Decompressor.java` should support this via the `Entry.Type type(ZipArchiveEntry e)` method. Let's hope it's a quick win
